charmcraft/reactive_plugin.py

```diff
@@ -0,0 +1,180 @@
+                        f"does not depend on a part named 'charm-tools'"
+                    ),
+                ) from err
+
+
+class ReactivePlugin(plugins.Plugin):
+    """Build a reactive charm using charm-tools."""
+
+    properties_class = ReactivePluginProperties
+    validator_class = ReactivePluginEnvironmentValidator
+
+    @classmethod
+    def get_build_snaps(cls) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        command = [
+            sys.executable,
+            "-I",
+            __file__,
+            self._part_info.project_name,
+            str(self._part_info.part_build_dir),
+            str(self._part_info.part_install_dir),
+        ]
+
+        return [" ".join(shlex.quote(i) for i in command)]
+
+
+def build(*, charm_name: str, build_dir: Path, install_dir: Path) -> None:
+    """Build a charm using charm tool.
+
+    The charm tool is used to build reactive charms, the build process
+    is as follows:
+
+    - Remove any charmcraft.yaml from the build directory (occurs for
+      local in-tree builds)
+
+    - Run charm proof to ensure the charm
+      would build with no errors (warnings are allowed) - - Create a
+      symlink to where charm
+
+    - Link charm tool's build directory to the part lifecycle's
+      install_dir.
+
+    - Run "charm build"
+    """
+    # Remove the charmcraft.yaml so it is not primed for in-tree builds.
+    charmcraft_yaml = build_dir / "charmcraft.yaml"
+    if charmcraft_yaml.exists():
+        charmcraft_yaml.unlink()
+
+    # Verify the charm is ok from a charm tool point of view.
+    try:
+        subprocess.run(["charm", "proof"], check=True)
+    except subprocess.CalledProcessError as call_error:
+        if call_error.returncode >= 200:
+            raise call_error
+
+    # Link the installation directory to the place where charm creates
+    # the charm.
+    charm_build_dir = build_dir / charm_name
+    if not charm_build_dir.exists():
+        charm_build_dir.symlink_to(install_dir, target_is_directory=True)
+
+    try:
+        subprocess.run(["charm", "build", "-o", build_dir], check=True)
+    except subprocess.CalledProcessError as call_error:
+        if call_error.returncode >= 200:
+            raise call_error
+    finally:
+        charm_build_dir.unlink()
+
+
+if __name__ == "__main__":
+    try:
+        build(charm_name=sys.argv[1], build_dir=Path(sys.argv[2]), install_dir=Path(sys.argv[3]))
+    except subprocess.CalledProcessError as call_error:
```

Comment:
Don't leak `CalledProcessError` and the magic number `200` out of `build`.

It's get simpler, even: make `build` to return the return code (see comments above)